### PR TITLE
Rewrote network stack

### PR DESCRIPTION
**Pull request by Zegeger**
Rewrote network stack to support buffering messages and then dispatching as 1 or more combined or split packets.  EncryptedChecksum seems to work across the board.  Sending Packets with fragments from different groups does not seem to cause an issue.
Packets are queued and sent based on calls from the Engine.  Configurable delay (5ms currently)
Reworked CachedPackets some, and now clear the list based on AckSeq packets from client.  Sending AckSeq packets to the client seems to cause issues currently, need more testing.
Moved some things around to fit this new model.
Put message Group as a property on each message. And send this in fragment.  Sounds now work again.
Created new packet classes to encapsulate data in AuthenticationManager and CharacterManager and allow them to be send through the same model (though packets will never be manipulated).

### CODE REVIEW

**Comment by Mag-nus** [comment]
GameMessageF7E5 should be named GameMessageDDD_Interrogation

There's a lot of hard coded group types still: 0x5, 0x9, 0xA, etc.. We should use an enum for the groups.

NetworkSession.cs, and anywhere else, should use DateTime.UtcNow instead of DateTime.Now

This is a big PR so it will require further review.. I like what I see so far though, can't wait to get this merged.. I'll have to look at it a bit more later, and will want Mogwai to look at it as well.

File: Source/ACE/Network/NetworkSession.cs [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+        public SessionConnectionData ConnectionData { get; private set; }
+
+        public NetworkSession(Session session, ConnectionType connType)
+        {
+            this.session = session;
+            this.connectionType = connType;
+            ConnectionData  = new SessionConnectionData(connType);
+            currentBundle = new NetworkBundle();
+            nextSend = DateTime.Now;
```
Comment (Mogwai-AC) [inline Source/ACE/Network/NetworkSession.cs]: UtcNow or daylight savings time will crash the server.

File: Source/ACE/Network/GameEvent/Events/GameEventDisplayParameterizedStatusMessage.cs [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
-using ACE.Entity.Enum;
+﻿using ACE.Entity.Enum;
 
 namespace ACE.Network.GameEvent.Events
 {
     public class GameEventDisplayParameterizedStatusMessage : GameEventMessage
     {
-        public GameEventDisplayParameterizedStatusMessage(Session session, StatusMessageType2 statusMessageType2, string message) : base(GameEventType.DisplayParameterizedStatusMessage, session)
+        public GameEventDisplayParameterizedStatusMessage(Session session, StatusMessageType2 statusMessageType2, string message) 
+            : base(GameEventType.DisplayParameterizedStatusMessage, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventDisplayParameterizedStatusMessage.cs]: put the 0x9 in a enum

File: Source/ACE/Network/GameEvent/Events/GameEventDisplayStatusMessage.cs [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
-using ACE.Entity.Enum;
+﻿using ACE.Entity.Enum;
 
 namespace ACE.Network.GameEvent.Events
 {
     public class GameEventDisplayStatusMessage : GameEventMessage
     {
-        public GameEventDisplayStatusMessage(Session session, StatusMessageType1 statusMessageType1) : base(GameEventType.DisplayStatusMessage, session)
+        public GameEventDisplayStatusMessage(Session session, StatusMessageType1 statusMessageType1) 
+            : base(GameEventType.DisplayStatusMessage, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventDisplayStatusMessage.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs [hunk 4]
```diff
@@ -30,7 +30,7 @@ public enum FriendsUpdateTypeFlag
         /// </summary>
         /// <param name="session"></param>
         public GameEventFriendsListUpdate(Session session) 
-            : base (GameEventType.FriendsListUpdate, session)
+            : base (GameEventType.FriendsListUpdate, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventHouseStatus.cs [hunk 5]
```diff
@@ -1,9 +1,9 @@
-﻿
-namespace ACE.Network.GameEvent.Events
+﻿namespace ACE.Network.GameEvent.Events
 {
     public class GameEventHouseStatus : GameEventMessage
     {
-        public GameEventHouseStatus(Session session) : base(GameEventType.HouseStatus, session)
+        public GameEventHouseStatus(Session session)
+            : base(GameEventType.HouseStatus, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventHouseStatus.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventIdentifyObjectResponse.cs [hunk 6]
```diff
@@ -3,7 +3,7 @@ namespace ACE.Network.GameEvent.Events
 {
     public class GameEventIdentifyObjectResponse : GameEventMessage
     {
-        public GameEventIdentifyObjectResponse(Session session, uint objectID) : base(GameEventType.IdentifyObjectResponse, session)
+        public GameEventIdentifyObjectResponse(Session session, uint objectID) : base(GameEventType.IdentifyObjectResponse, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventIdentifyObjectResponse.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventPingResponse.cs [hunk 7]
```diff
@@ -1,9 +1,9 @@
-﻿
-namespace ACE.Network.GameEvent.Events
+﻿namespace ACE.Network.GameEvent.Events
 {
     public class GameEventPingResponse : GameEventMessage
     {
-        public GameEventPingResponse(Session session) : base(GameEventType.PingResponse, session) { }
+        public GameEventPingResponse(Session session) 
+            : base(GameEventType.PingResponse, 0x9, session) { }
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventPingResponse.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventPlayerDescription.cs [hunk 8]
```diff
@@ -60,7 +60,7 @@ private enum DescriptionOptionFlag
         }
 
         public GameEventPlayerDescription(Session session)
-            : base(GameEventType.PlayerDescription, session)
+            : base(GameEventType.PlayerDescription, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventPlayerDescription.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventPopupString.cs [hunk 9]
```diff
@@ -1,9 +1,9 @@
-﻿
-namespace ACE.Network.GameEvent.Events
+﻿namespace ACE.Network.GameEvent.Events
 {
     public class GameEventPopupString : GameEventMessage
     {
-        public GameEventPopupString(Session session, string message) : base(GameEventType.PopupString, session)
+        public GameEventPopupString(Session session, string message)
+            : base(GameEventType.PopupString, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventPopupString.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventSetTurbineChatChannels.cs [hunk 10]
```diff
@@ -7,7 +7,8 @@ public class GameEventSetTurbineChatChannels : GameEventMessage
         /// unknown1, unknown2 unknown3 do not appear to be used by the latest client.<para />
         /// They are possibly accessed by an admin client. They could be used in the future for admin chat via a decal plugin (with some sort of auth layer). 
         /// </summary>
-        public GameEventSetTurbineChatChannels(Session session, uint allegiance, uint general, uint trade, uint lfg, uint rolePlay, uint olhtoi, uint society, uint unknown1, uint unknown2, uint unknown3) : base(GameEventType.SetTurbineChatChannels, session)
+        public GameEventSetTurbineChatChannels(Session session, uint allegiance, uint general, uint trade, uint lfg, uint rolePlay, uint olhtoi, uint society, uint unknown1, uint unknown2, uint unknown3) 
+            : base(GameEventType.SetTurbineChatChannels, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventSetTurbineChatChannels.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventTell.cs [hunk 11]
```diff
@@ -1,11 +1,11 @@
...
-using ACE.Entity.Enum;
+﻿using ACE.Entity.Enum;
 
 namespace ACE.Network.GameEvent.Events
 {
     public class GameEventTell : GameEventMessage
     {
-        public GameEventTell(Session session, string messageText, string senderName, uint senderID, uint targetID, ChatMessageType chatMessageType) : base(GameEventType.Tell, session)
+        public GameEventTell(Session session, string messageText, string senderName, uint senderID, uint targetID, ChatMessageType chatMessageType) 
+            : base(GameEventType.Tell, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventTell.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameEvent/Events/GameEventUpdateTitle.cs [hunk 12]
```diff
@@ -1,9 +1,9 @@
-﻿
-namespace ACE.Network.GameEvent.Events
+﻿namespace ACE.Network.GameEvent.Events
 {
     public class GameEventUpdateTitle : GameEventMessage
     {
-        public GameEventUpdateTitle(Session session, uint title) : base(GameEventType.UpdateTitle, session)
+        public GameEventUpdateTitle(Session session, uint title) 
+            : base(GameEventType.UpdateTitle, 0x9, session)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventUpdateTitle.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameMessages/Messages/GameMessageCharacterCreateResponse.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
+﻿using ACE.Entity;
+using ACE.Entity.Enum;
+using ACE.Network.Enum;
+
+namespace ACE.Network.GameMessages.Messages
+{
+    public class GameMessageCharacterCreateResponse : GameMessage
+    {
+        public GameMessageCharacterCreateResponse(CharacterGenerationVerificationResponse response, ObjectGuid guid, string charName) 
+            : base(GameMessageOpcode.CharacterCreateResponse, 0x9)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameMessages/Messages/GameMessageCharacterCreateResponse.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameMessages/Messages/GameMessageCharacterDelete.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
+﻿using ACE.Entity;
+using ACE.Entity.Enum;
+using ACE.Network.Enum;
+
+namespace ACE.Network.GameMessages.Messages
+{
+    public class GameMessageCharacterDelete : GameMessage
+    {
+        public GameMessageCharacterDelete() 
+            : base(GameMessageOpcode.CharacterDelete, 0x9)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameMessages/Messages/GameMessageCharacterDelete.cs]: put 0x9 in an enum

File: Source/ACE/Network/GameMessages/Messages/GameMessageCreateObject.cs [hunk 15]
```diff
@@ -0,0 +1,14 @@
+﻿using ACE.Entity;
+using ACE.Entity.Enum;
+
+namespace ACE.Network.GameMessages.Messages
+{
+    public class GameMessageCreateObject : GameMessage
+    {
+        public GameMessageCreateObject(WorldObject worldObject) 
+            : base(GameMessageOpcode.ObjectCreate, 0xA)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameMessages/Messages/GameMessageCreateObject.cs]: put 0xA in an enum

File: Source/ACE/Network/GameMessages/Messages/GameMessageF7E5.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
+﻿using ACE.Entity.Enum;
+
+namespace ACE.Network.GameMessages.Messages
+{
+    public class GameMessageF7E5: GameMessage
+    {
+        public GameMessageF7E5()
+            : base(GameMessageOpcode.DDD_Interrogation, 0x5)
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameMessages/Messages/GameMessageF7E5.cs]: put 0x5 in an enum

File: Source/ACE/Network/NetworkSession.cs (hunk 1 above)
Comment (Mag-nus) [inline Source/ACE/Network/NetworkSession.cs]: I did a global replace for Every DateTime.Now with DateTime.UtcNow and it seems to work ok.

**Comment by Mag-nus** [comment]
Another thing I'm noticing is that if you connect, close the client with the X when you're in portal space, and reconnect again.. repeat this a few times, the client will eventually crash, and the server seems to stop listening to new connections. 

Even after the server forcibly removes the clients, new connections are still not being accepted.

his happens with DateTime.Now and DateTime.UtcNow

I am unable to reproduce these steps on the current main branch.

**Comment by Zegeger** [comment]
Applied requested changes and I believe fixed the crash issues.

**Pull request merged**